File: packages/economic_agents/src/economic_agents/interfaces/wallet.py
"""Wallet interface for financial operations."""

from abc import ABC, abstractmethod
from dataclasses import dataclass
from datetime import datetime
from typing import List


@dataclass
class Transaction:
    """Represents a financial transaction."""

    tx_id: str
    from_address: str
    to_address: str
    amount: float
    timestamp: datetime
    status: str  # "pending", "confirmed", "failed"
    memo: str


class WalletInterface(ABC):
    """Abstract interface for wallet operations."""

    @abstractmethod
    async def get_balance(self) -> float:
        """Returns current wallet balance."""

    @abstractmethod
    async def send_payment(self, to_address: str, amount: float, memo: str = "") -> Transaction:
        """Sends payment to specified address."""

    @abstractmethod
    async def get_address(self) -> str:
        """Returns wallet's receiving address."""

    @abstractmethod
    async def get_transaction_history(self, limit: int = 100) -> List[Transaction]:
        """Returns recent transactions."""

    @abstractmethod
    async def receive_payment(self, from_address: str, amount: float, memo: str = "") -> Transaction:
        """Records incoming payment and updates balance.

        In real implementations, this would be triggered by blockchain events.
        In simulation, this is called directly when tasks are completed.

        Args:
            from_address: Address funds are coming from
            amount: Payment amount (must be positive)
            memo: Optional transaction memo

        Returns:
            Transaction record for the incoming payment

        Raises:
            ValueError: If amount is not positive
        """

File: packages/economic_agents/crates/economic-agents-interfaces/src/wallet.rs
//! Wallet interface for payment operations.

use async_trait::async_trait;

use crate::{Currency, Result, Transaction};

/// Interface for wallet/payment operations.
///
/// Implementations may connect to real cryptocurrency wallets,
/// mock in-memory wallets, or API-based payment services.
#[async_trait]
pub trait Wallet: Send + Sync {
    /// Get the current balance.
    async fn get_balance(&self) -> Result<Currency>;

    /// Get the wallet address.
    async fn get_address(&self) -> Result<String>;

    /// Send a payment to another address.
    ///
    /// # Arguments
    /// * `to` - Destination address
    /// * `amount` - Amount to send
    /// * `memo` - Optional transaction memo
    ///
    /// # Returns
    /// The transaction record on success.
    async fn send_payment(
        &self,
        to: &str,
        amount: Currency,
        memo: Option<&str>,
    ) -> Result<Transaction>;

    /// Receive a payment (credit the wallet).
    ///
    /// # Arguments
    /// * `from` - Source address (or None for deposits)
    /// * `amount` - Amount to receive
    /// * `memo` - Optional transaction memo
    ///
    /// # Returns
    /// The transaction record on success.
    async fn receive_payment(
        &self,
        from: Option<&str>,
        amount: Currency,
        memo: Option<&str>,
    ) -> Result<Transaction>;

    /// Get transaction history.
    ///
    /// # Arguments
    /// * `limit` - Maximum number of transactions to return
    ///
    /// # Returns
    /// List of transactions, most recent first.
    async fn get_transaction_history(&self, limit: usize) -> Result<Vec<Transaction>>;
}
